Rework test setup to use tctildr and enable libtpms

joholl opened the pull request:
Rework integration test framework with the goal of using an arbitrary tcti via tctildr and enabling tcti-libtpms. This should simplify the test framework, modernize the tcti handling and speed up testing.

Additionally, I turned the sapi/esapi tests into standalone executables (without need for additional processes startup, dumpstate, transientempty, ...). However, I could not pull that off for FAPI, mainly because of the EK logic in `fint-log-compiler.sh`. That means, for now I could not remove `test/helper/*`.

As standalone processes, the sapi/esapi tests can be tested against libtpms without any state file. The fapi tests however require a state file (but do indeed run green).

Unfortunately, small incremental changes are not really possible here, so please excuse the massive commit. I tried to clean up in the process, therefore the streamlined setup/teardown routines for sapi/esapi and the deletion of the old equivalent.

Feedback welcome! (Please don't make me work weeks on this, I spent already much more time that I anticipated)

JuergenReppSIT commented:
A first remark: I have tried to run the tests with physical TPMs as described in the README. There were two fails. In these two cases the error whether the TPM command is implemented was checked only with:
`r == TPM2_RC_COMMAND_CODE` and the resource manager layer was not checked.
The README describes the configure option `--with-device=/dev/tpm0` but  `/dev/tpmrm0` was used in the integration tests. Therefore these fails did occur.

joholl commented:
Good catch, I fixed that.

The problem is that we had both `INTEGRATION_TCTI` and `INTEGRATION_ARGS` and I didn't consider the latter.  I removed `INTEGRATION_ARGS` completely.

Let me bring up a related topic. At the moment we have an autoconf-based auto-detect for tctis for testing (which does not consider libtpms) and a manual configure switch `--with-device`.

What I'd like to see in a future PR is support for libtpms in the auto-detect and more importantly, a configure switch `--with-integration-tcti` which enables the user to choose any tcti used for testing.

https://github.com/tpm2-software/tpm2-tss/blob/11d9d056fb3b559646eaa88995a4db822fa8af3b/configure.ac#L445-L470

joholl commented:
@AndreasFuchsTPM @JuergenReppSIT @wxleong ping

wxleong commented:
The rest looks fine to me, provided that `Esys_SetTimeout()` is not a concern.

joholl commented:
@wxleong The timeout is handed down to the tcti receive call and ignored by our most tctis. For tcti-device, receive might return `TRY_AGAIN` which should be handled gracefully by ESAPI.

joholl commented on test/integration/main-esys.c:
```diff
@@ -184,69 +36,35 @@ tcti_proxy_initialize(
+        return rc;
     }
 
-    rc = Esys_SetTimeout(esys_context, TSS2_TCTI_TIMEOUT_BLOCK);
```
As mentioned above: the timeout is handed down to the tcti receive call and ignored by our most tctis. For tcti-device, receive might return `TRY_AGAIN` which should be handled gracefully by ESAPI.
